File: euler/numbertheory.go
package euler

var primepilist [primeTableLength]int64

//number of primes less than or equal to n
func PrimePi(n int64) int64 {
	if n < 2 {
		return 0
	}

	if n == 2 {
		return 1
	}

	if n < primeTableLength && primepilist[n] != 0 {
		return primepilist[n]
	}

	var answer int64
	if IsPrime(n) {
		answer = 1 + PrimePi(n-1)
		if answer < primeTableLength {
			primeTable[answer] = n
		}
	} else {
		answer = PrimePi(n - 1)

	}

	if n < primeTableLength {
		primepilist[n] = answer
	}
	return answer
}

var factorialtable = make(map[int64]int64)

func Factorial(n int64) int64 {
	if n == 0 {
		return 1
	}
	if answer, ok := factorialtable[n]; ok {
		return answer
	}

	answer := Factorial(n-1) * n

	factorialtable[n] = answer

	return answer
}

File: cpp/number_theory.cc
#include <vector>

using std::vector;

bool is_prime(int n) {
  if (n <= 1) {
    return false;
  }
  for (int a = 2; a * a <= n; a++) {
    if (n % a == 0) {
      return false;
    }
  }
  return true;
}

vector<int> primes_below(int n) {
  vector<int> answer;
  int x = 2;

  while (x < n) {
    if (is_prime(x)) {
      answer.push_back(x);
    }
    x++;
  }

  return answer;
}
